chore: Adding wrappers for guild and privileged commands

plamoni opened the pull request:
Consolidates logic for checking whether a command is executed within a guild and also checking for the required permissions for the command. This simplifies command implementation and reduces the opportunity for errors.

Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a [descriptive title](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=prepare-a-good-pr-title) (**not** a vague title like `Update index.md`)

This PR is as much a proposal as anything. It has not yet been thoroughly tested (will perform more testing if the approach is approved).

plamoni committed:
chore: Adding wrappers for guild and privileged commands

Consolidates logic for checking whether a command is executed within a guild and also checking for the required permissions for the command. This simplifies command implementation and reduces the opportunity for errors.

bradtaniguchi reviewed (commented):
I'd add some comments to at least the high level types/functions added to help describe how to use it. Specifically the `wrapCommands` function, which has a name that explains what it does, but not why or what a "wrapped" command represents. 

Also what's the purpose of a "wrapped command", which is the default case when a command isn't a privileged or guild command?

plamoni committed:
fix: lint and deferred interaction reply bug

plamoni committed:
fix: removing redundant permission check on privateChannel.ts

naomi-lgbt commented on src/utils/wrapCommands.ts:
```diff
@@ -0,0 +1,89 @@
+
+const isGuildCommand = (command: UnwrappedCommand): command is GuildCommand => {
+  return !!(command as GuildCommand).guildOnly;
+};
```
```suggestion
const isPrivilegedCommand = (
  command: UnwrappedCommand
): command is PrivilegedCommand =>
  isGuildCommand(command) && "requiredPermissions" in command;

const isGuildCommand = (command: UnwrappedCommand): command is GuildCommand =>
  "guildOnly" in command;
```

Rather than coercing, we could check for the key `in` object?

plamoni replied:
The reason I went with (and tend to lean towards) the use of `as` like this is because of [property renaming](https://developers.google.com/closure/compiler/docs/limitations#implications-of-global-variable,-function,-and-property-renaming:). While I realize we don't use  it for this project, I've come to consider it a best practice to write code that will not break when its enabled. By mixing string references and dot references to the property, your suggested approach would likely break when subjected to property renaming.

There's [no runtime performance advantage](https://jsbench.me/19lcurvvet/1) to either approach:

![image](https://user-images.githubusercontent.com/306088/212377782-00a146d1-9f61-44d6-a693-1cdc6d40bdf2.png)

However, there _is_ a logical difference between the two approaches that's notable:

```js
const myObj = { prop: undefined };
console.log("prop" in myObj); // true
console.log(!!myObj.prop); // false
```

In the case of my code above, the type guard is doing double duty: checking both the presence _and_ truthiness of the target property. Due to how [`GuildCommand` is defined](https://github.com/plamoni/discord-bot/blob/5aa7b646f0562fce6c2c5aa327b62c7114f36964/src/interfaces/GuildCommand.ts#L6), an object with the `guildOnly` property set to anything other than `true` does not satisfy the `GuildCommand` interface. Therefore, a type guard checking only for the _presence_ of the property risks it mis-identifying a command with `guildOnly: false` as  being a `GuildCommand`.

To be fair, my solution does identify any truthy value in `guildOnly` as indicating satisfaction of the `GuildCommand` interface, though I think it's fair to say someone putting a truthy value in `guildOnly` reasonably intends it to be a guild-only command, while someone putting `false` in that field would be (with reason) surprised to find it was interpreted as a guild-only command.

Given that we've spent a combined several minutes on this one bit of code, I think it's only fair that I take _some_ action on it. Why not make it more resilient to runtime issues and more specifically accurate? e.g.:

```ts
const isGuildCommand = (command: UnwrappedCommand): command is GuildCommand =>
  (command as GuildCommand)?.guildOnly === true;
```

This has the advantage of being resilient to being passed a non-object argument (something that really could only happen by circumventing type checking) and also explicitly checking for a value of `true` instead of accepting any truthy value. It is also resilient to property renaming and omits the unnecessary block wrapper.

Let me know what you think and I'll make the change for both guards.

naomi-lgbt replied:
Thanks for the detailed writeup - this makes quite a bit of sense.

I'd go ahead with the more resilient solution you proposed.

plamoni replied:
Done.

naomi-lgbt commented on src/utils/wrapCommands.ts:
```diff
@@ -0,0 +1,89 @@
+      return wrapPrivilegedCommand(command);
+    } else if (isGuildCommand(command)) {
+      return wrapGuildCommand(command);
+    }
```
```suggestion
    if (isPrivilegedCommand(command)) {
      return wrapPrivilegedCommand(command);
    }
    if (isGuildCommand(command)) {
      return wrapGuildCommand(command);
    }
```

Nitpicking, but don't need `else if` on an early return.

naomi-lgbt reviewed (changes requested):
Some nitpicking

plamoni committed:
Update src/utils/wrapCommands.ts

Removing unnecessary else-if.

Co-authored-by: Naomi Carrigan <nhcarrigan@gmail.com>

plamoni committed:
Merge branch 'freeCodeCamp:main' into command-perms

plamoni committed:
chore: Updating isGuildCommand logic to be more precise

Per discussion in #261, making sure that guildOnly is true and not just
truthy.

plamoni committed:
chore: Adding more comments

Per discussion in #261

plamoni committed:
Merge branch 'main' into command-perms

naomi-lgbt reviewed (approved):
Okay, I've given this some testing, and I think we're good to go. Thanks for the hard work on this one!
